src/main/java/edu/umich/its/spe/SPEMaster.java

```diff
@@ -0,0 +1,378 @@
+
+	//{"putPlcExamScoreResponse":{"putPlcExamScoreResponse":{"Status":"SUCCESS","Form":7,"ID":"abc"
+
+	/*********** format assignments from the data warehouse ********/
+	
+	// After getting the JSON from the data warehouse format the SPE grades.
+	protected ArrayList<HashMap<String,String>> convertSPEGradesFromDataWarehouseJSON(String assignmentJSON) {
+	
+		M_log.debug("cSGFDWJ: assignmentJSON: {}",assignmentJSON);
+		// if nothing in it then return an empty list.
+		if (assignmentJSON == null || assignmentJSON.length() <=2) {
+			ArrayList<HashMap<String,String>> emptyArrayList = new ArrayList<HashMap<String,String>>();
+			return emptyArrayList;
+		}
+	
+		JSONArray assignments = parseCanvasAssignmentJSON(assignmentJSON);
+		ArrayList<HashMap<String,String>> gradeMapList = convertAssignmentsToGradeMaps(assignments);
+		M_log.debug("cSGFDWJ: gradeMaplist: {}",gradeMapList);
+		return gradeMapList;
+	}
+
+	// convert the list of DataWarehouse assignments to a list of grade maps for MPathways.
+	protected ArrayList<HashMap<String, String>> convertAssignmentsToGradeMaps(JSONArray canvasAssignments) {
+		ArrayList<HashMap<String,String>> gradeMapList = new ArrayList<HashMap<String, String>>();
+	
+		// JSONArray is not iterable so can't use the nice "for each" syntax.
+		for(int i = 0; i<canvasAssignments.length(); i++) {
+			JSONObject assignment = canvasAssignments.getJSONObject(i);
+			HashMap<String, String> grademap = convertAssignmentToGradeMap(assignment);
+			gradeMapList.add(grademap);
+		}
+	
+		return gradeMapList;
+	}
+
+	// convert a single JSON version of an assignment to a grademap.
+	static protected HashMap<String, String> convertAssignmentToGradeMap(JSONObject assignment) throws JSONException {
+		HashMap<String,String> grademap = new HashMap<String,String>();
+		// Score may be read as a number instead of a string so pull out as an object and convert to a string.
+		grademap.put("Score",JSONObject.valueToString(assignment.get("Score")));
+		grademap.put("Unique_Name", assignment.getString("Unique_Name"));
+		return grademap;
+	}
+
+	// Take the assignment JSON string from the data warehouse and return the array of grades as a parsed JSON array.
+	static protected JSONArray parseCanvasAssignmentJSON(String gradeJSON) throws JSONException {
+		JSONObject jo = new JSONObject(gradeJSON);
+		JSONObject jAI = jo.getJSONObject("AssignmentInfo");
+	
+		// Get the data as a JSON array.
+		JSONArray jAD = jAI.optJSONArray("AssignmentData");
+	
+		// If there is a single grade it may appear as a single object, so deal with that.
+		if (jAD == null) {
+			jAD = new JSONArray();
+			jAD.put(jAI.getJSONObject("AssignmentData"));
```

Comment:
I'll defer to the "Rule of 3" and make constants if there are any more references to it.